nbs/99_manuscript/intro/py/05-relationships_analysis.py

```diff
@@ -0,0 +1,313 @@
+            "x": x,
+            "y": y,
+        }
+    ),
+    ignore_index=True,
+)
+
+# %% [markdown] tags=[]
+# Create a dictionary with easier access to datasets
+
+# %% tags=[]
+datasets = {
+    idx: df.drop(columns="dataset") for idx, df in datasets_df.groupby("dataset")
+}
+
+
+# %% [markdown] tags=[]
+# # Plot
+
+# %% tags=[]
+def get_cm_line_points(x, y, max_parts, parts):
+    """
+    Given two data vectors (x and y) and the max_parts and parts
+    returned from calling cm, this function returns two arrays with
+    scalars to draw the lines that separates clusters in x and y.
+    """
+    # get the clustermatch partitions that maximize the coefficient
+    x_max_part = parts[0][max_parts[0]]
+    x_unique_k = {}
+    for k in np.unique(x_max_part):
+        data = x[x_max_part == k]
+        x_unique_k[k] = data.min(), data.max()
+    x_unique_k = sorted(x_unique_k.items(), key=lambda x: x[1][0])
+
+    y_max_part = parts[1][max_parts[1]]
+    y_unique_k = {}
+    for k in np.unique(y_max_part):
+        data = y[y_max_part == k]
+        y_unique_k[k] = data.min(), data.max()
+    y_unique_k = sorted(y_unique_k.items(), key=lambda x: x[1][0])
+
+    x_line_points, y_line_points = [], []
+
+    for idx in range(len(x_unique_k) - 1):
+        k, (k_min, k_max) = x_unique_k[idx]
+        nk, (nk_min, nk_max) = x_unique_k[idx + 1]
+
+        x_line_points.append((k_max + nk_min) / 2.0)
+
+    for idx in range(len(y_unique_k) - 1):
+        k, (k_min, k_max) = y_unique_k[idx]
+        nk, (nk_min, nk_max) = y_unique_k[idx + 1]
+
+        y_line_points.append((k_max + nk_min) / 2.0)
+
+    return x_line_points, y_line_points
+
+
+# %% tags=[]
+with sns.plotting_context("paper", font_scale=1.8):
+    g = sns.FacetGrid(data=datasets_df, col="dataset", col_wrap=4, height=5)
+    g.map(sns.scatterplot, "x", "y", s=50, alpha=1)
+    g.set_titles(row_template="{row_name}", col_template="{col_name}")
+
+    for ds, ax in g.axes_dict.items():
+        df = datasets[ds].to_numpy()
+        x, y = df[:, 0], df[:, 1]
+
+        # pearson and spearman
+        r = pearsonr(x, y)[0]
+        rs = spearmanr(x, y)[0]
+
+        # clustermatch
+        c, max_parts, parts = cm(x, y, return_parts=True)
+        c = cm(x, y)
+
+        x_line_points, y_line_points = get_cm_line_points(x, y, max_parts, parts)
+        for yp in y_line_points:
+            ax.hlines(y=yp, xmin=-0.5, xmax=20, color="r", alpha=0.5)
+
+        for xp in x_line_points:
+            ax.vlines(x=xp, ymin=1.5, ymax=14, color="r", alpha=0.5)
+
+        # add text box for the statistics
+        stats = f"$r$ = {r:.2f}\n" f"$r_s$ = {rs:.2f}\n" f"$c$ = {c:.2f}"
+        bbox = dict(boxstyle="round", fc="white", ec="black", alpha=0.75)
+        ax.text(
+            0.95,
+            0.07,
+            stats,
+            fontsize=14,
+            bbox=bbox,
+            transform=ax.transAxes,
+            horizontalalignment="right",
+        )
+
+    plt.savefig(
```

Comment:
This is SUCH a cool plot! Its so awesome to see how clustermatch is able to detect the more complex and nonlinear signals in the data compared to Pearson and Spearman. It also is insensitive to outliers too.

I wonder if you tried varying the number of samples you have if clustermatch will still be able to detect this pattern, just because these patterns are very clean right now.

Reply:
Have you tried this on a subset of real data? Just curious

Reply:
Thank you, Alex! We did a more comprehensive analysis of Clustermatch [here](https://academic.oup.com/bioinformatics/article/35/11/1931/5144171), but we only vary the number of samples for time complexity analysis. I would expect that, if you have fewer samples, the coefficient would behave in a more unstable way, for example, if you generate random data with 10, 20, 30 samples for each variable. But I think that's what all coefficients do with less data.

In that paper we did run it on real data also, but did not compare this with other coefficients. We are doing that in this paper with gene expression data.